Web assert add timeout

alexisLefebvre opened the pull request:
May be a solution for:

    https://github.com/minkphp/Mink/issues/606

I worked with @lenybernard in order to merge his changes to the current version of Mink, see the previous discussion for reference:

    https://github.com/Behat/MinkExtension/pull/210

(I put links in blockquote to avoid creating back links on these pages)

Commit:
implements waitFor for each WebAssert correct assertions

Charlie-Lucas committed:
increase timeout to 10 seconds

Commit:
Merge pull request #1 from Charlie-Lucas/patch-1

increase timeout to 10 seconds

Commit:
add timeout for address(Not)Equals function

alexisLefebvre committed:
Merge branch 'master' of github.com:lenybernard/Mink into WebAssert-add-timeout

alexisLefebvre commented on src/WebAssert.php:
```diff
@@ -46,12 +47,17 @@ public function __construct(Session $session)
 
-        $this->assert($actual === $expected, sprintf('Current page is "%s", but "%s" expected.', $actual, $expected));
+        $this->assert(
+            $this->session->getPage()->waitFor($timeout, function () use ($cleanedUrl, &$actual) {
```
@lenybernard Do you think that we should keep the `waitFor()` for the address? I'm pretty sure that the address of a page is not supposed to change once a page if loaded (if Mink wait for redirections, I didn't check that).

lenybernard replied:
Actually, the address could change with javascript so I guess it can be useful

alexisLefebvre commented on src/WebAssert.php:
```diff
@@ -61,12 +67,19 @@ public function addressEquals($page)
+                $actual = $this->getCurrentUrlPath();
+                // TODO Don't fail on first try if address are different.
+                // TODO This condition is true and break the loop.
+                return $actual !== $cleanedUrl;
```
@lenybernard I think that this condition can't work.

Let's say that the test expect the address to be different from `path/`:
- if the URL is `path/` during one second, then `/file`, the test will work (after trying during one second)
- if the URL is `/file` during one second, then `path/`, the test will work, the condition will be evaluated to true during the first loop iteration, it won't wait several seconds, so the test will be marked as passed incorrectly

lenybernard reviewed (changes requested):
Please run php-cs-fixer before

lenybernard commented on src/WebAssert.php:
```diff
@@ -28,6 +28,7 @@
 class WebAssert
 {
     protected $session;
+    const timeout = 5;
```
constant should be uppercased please

alexisLefebvre replied:
This must be documented in CHANGES.md and in README.md.

lenybernard commented on src/WebAssert.php:
```diff
@@ -61,12 +67,19 @@ public function addressEquals($page)
+        $this->assert(
+            $this->session->getPage()->waitFor($timeout, function () use ($cleanedUrl, &$actual) {
+                $actual = $this->getCurrentUrlPath();
+                // TODO Don't fail on first try if address are different.
```
remove comments please

lenybernard commented on src/WebAssert.php:
```diff
@@ -378,22 +418,30 @@ public function responseNotMatches($regex)
             '%d %s found on the page, but should be %d.',
             count($nodes),
-            $this->getMatchingElementRepresentation($selectorType, $selector, count($nodes) !== 1),
+            $this->getMatchingElementRepresentation($selectorType, $selector, count($nodes) > 1),
```
why ?

lenybernard commented on src/WebAssert.php:
```diff
@@ -402,17 +450,23 @@ public function elementsCount($selectorType, $selector, $count, ElementInterface
      * @param string           $selectorType element selector type (css, xpath)
      * @param string|array     $selector     element selector
      * @param ElementInterface $container    document to check against
+     * @param integer          $timeout      time to wait until fail
```
prefer float like waitFor method

lenybernard commented on tests/WebAssertTest.php:
```diff
@@ -5,6 +5,26 @@
 use Behat\Mink\Exception\ExpectationException;
 use Behat\Mink\WebAssert;
 
+/**
```
don't add these comments

lenybernard commented on tests/WebAssertTest.php:
```diff
@@ -15,6 +35,7 @@ class WebAssertTest extends \PHPUnit_Framework_TestCase
      * @var WebAssert
      */
     private $assert;
+    const timeoutLimit = 0.11;
```
explane here wtf is this value 🤓

lenybernard replied:
uppercase

lenybernard reviewed (changes requested):
please add also some kind of docs in `CHANGES.md` and add a use case in the `README.md` to show off what we can do now with this amazing timeout and how we can configure it

Commit:
:white_check_mark: Add timeouts to WebAssert: Fix tests

Updates CHANGES.md
Set timeout as constant and decrease to 5 seconds
Fix deprecated argument passed to ElementNotFoundException()
